File: packages/core/src/providers/github.ts
import type { UpdateManifest, UpdateProvider } from '../types';
import { getCurrentPlatform } from '../utils/platform';

export interface GitHubProviderConfig {
  owner: string;
  repo: string;
  manifestAssetName?: string;
  /**
   * If true, falls back to the GitHub Releases API when the manifest
   * asset is not found.
   * @default true
   */
  fallbackToReleaseApi?: boolean;
}

interface GitHubReleaseAsset {
  name: string;
  browser_download_url: string;
}

interface GitHubReleaseResponse {
  tag_name: string;
  published_at: string;
  body: string;
  assets: GitHubReleaseAsset[];
}

/**
 * Creates an update provider backed by GitHub Releases.
 *
 * It tries to download a manifest asset first and, if not found,
 * optionally falls back to the Releases API.
 */
export function createGitHubProvider(config: GitHubProviderConfig): UpdateProvider {
  const {
    owner,
    repo,
    manifestAssetName = 'update-manifest.json',
    fallbackToReleaseApi = true
  } = config;

  const manifestUrl = `https://github.com/${owner}/${repo}/releases/latest/download/${manifestAssetName}`;
  const apiUrl = `https://api.github.com/repos/${owner}/${repo}/releases/latest`;

  const provider: UpdateProvider & {
    __tauriParams?: {
      github_owner: string;
      github_repo: string;
      manifest_url?: string;
    };
  } = {
    type: 'github',
    async fetchManifest(): Promise<UpdateManifest> {
      // First try manifest asset.
      try {
        const res = await fetch(manifestUrl);
        if (res.ok) {
          const data = (await res.json()) as UpdateManifest;
          return data;
        }
        if (!fallbackToReleaseApi || res.status !== 404) {
          throw new Error(`GitHub manifest request failed with status ${res.status}`);
        }
      } catch (err) {
        if (!fallbackToReleaseApi) {
          throw err instanceof Error ? err : new Error('Failed to fetch GitHub manifest');
        }
      }

      // Fallback to Releases API.
      try {
        const res = await fetch(apiUrl, {
          headers: {
            Accept: 'application/vnd.github+json'
          }
        });
        if (!res.ok) {
          throw new Error(`GitHub releases API failed with status ${res.status}`);
        }

        const release = (await res.json()) as GitHubReleaseResponse;
        const platform = await getCurrentPlatform();

        let url = '';
        if (platform) {
          const asset = selectAssetForPlatform(release.assets, platform);
          if (asset) {
            url = asset.browser_download_url;
          }
        }

        const manifest: UpdateManifest = {
          version: release.tag_name,
          pub_date: release.published_at,
          notes: release.body ?? '',
          platforms: url
            ? {
                [platform]: {
                  url
                }
              }
            : {}
        };

        return manifest;
      } catch (error) {
        throw error instanceof Error ? error : new Error('Failed to fetch GitHub release');
      }
    }
  };

  provider.__tauriParams = {
    github_owner: owner,
    github_repo: repo,
    manifest_url: manifestUrl
  };

  return provider;
}

function selectAssetForPlatform(
  assets: GitHubReleaseAsset[],
  platform: string
): GitHubReleaseAsset | undefined {
  if (!platform) return undefined;

  const lowerPlatform = platform.toLowerCase();

  return assets.find((asset) => {
    const name = asset.name.toLowerCase();
    if (lowerPlatform === 'windows-x86_64') {
      return name.endsWith('.exe') || name.endsWith('.msi');
    }
    if (lowerPlatform === 'darwin-x86_64') {
      return name.endsWith('.dmg') && !name.includes('arm64') && !name.includes('aarch64');
    }
    if (lowerPlatform === 'darwin-aarch64') {
      return name.endsWith('.dmg') && (name.includes('arm64') || name.includes('aarch64'));
    }
    if (lowerPlatform === 'linux-x86_64') {
      return (
        name.endsWith('.appimage') ||
        name.endsWith('.AppImage') ||
        name.endsWith('.deb') ||
        name.endsWith('.rpm')
      );
    }
    return false;
  });
}

File: crates/tauri-easy-updater/src/providers/github.rs
use reqwest::blocking::Client;
use reqwest::Url;
use serde::Deserialize;

use crate::types::{PlatformInfo, UpdateManifest};

#[derive(Debug, Deserialize)]
struct GitHubAsset {
    name: String,
    browser_download_url: String,
}

#[derive(Debug, Deserialize)]
struct GitHubRelease {
    tag_name: String,
    published_at: String,
    body: String,
    assets: Vec<GitHubAsset>,
}

pub fn fetch_manifest_from_github(
    owner: &str,
    repo: &str,
    manifest_url: &str,
) -> Result<UpdateManifest, String> {
    let client = Client::builder()
        .user_agent("tauri-easy-updater")
        .build()
        .map_err(|e| e.to_string())?;

    // First try manifest asset URL
    if !manifest_url.is_empty() {
        if let Ok(url) = Url::parse(manifest_url) {
            if let Ok(resp) = client.get(url).send() {
                if resp.status().is_success() {
                    let manifest = resp
                        .json::<UpdateManifest>()
                        .map_err(|e| format!("Failed to parse manifest: {e}"))?;
                    return Ok(manifest);
                }
            }
        }
    }

    // Fallback to GitHub Releases API
    let api_url = format!(
        "https://api.github.com/repos/{owner}/{repo}/releases/latest",
    );

    let resp = client
        .get(api_url)
        .send()
        .map_err(|e| format!("GitHub API request failed: {e}"))?;

    if !resp.status().is_success() {
        return Err(format!("GitHub API error: {}", resp.status()));
    }

    let release = resp
        .json::<GitHubRelease>()
        .map_err(|e| format!("Failed to parse GitHub release: {e}"))?;

    let platform_key = current_platform_key();
    let mut platforms = std::collections::HashMap::new();

    if let Some(asset) = select_asset_for_platform(&release.assets, &platform_key) {
        platforms.insert(
            platform_key.clone(),
            PlatformInfo {
                url: asset.browser_download_url.clone(),
            },
        );
    }

    Ok(UpdateManifest {
        version: release.tag_name,
        pub_date: release.published_at,
        notes: release.body,
        platforms,
    })
}

fn current_platform_key() -> String {
    let os = match std::env::consts::OS {
        "windows" => "windows",
        "macos" => "darwin",
        "linux" => "linux",
        other => other,
    };

    let arch = match std::env::consts::ARCH {
        "x86_64" => "x86_64",
        "aarch64" => "aarch64",
        other => other,
    };

    format!("{os}-{arch}")
}

fn select_asset_for_platform<'a>(
    assets: &'a [GitHubAsset],
    platform: &str,
) -> Option<&'a GitHubAsset> {
    let platform = platform.to_lowercase();

    assets.iter().find(|asset| {
        let name = asset.name.to_lowercase();
        if platform == "windows-x86_64" {
            name.ends_with(".exe") || name.ends_with(".msi")
        } else if platform == "darwin-x86_64" {
            name.ends_with(".dmg") && !name.contains("arm64") && !name.contains("aarch64")
        } else if platform == "darwin-aarch64" {
            name.ends_with(".dmg") && (name.contains("arm64") || name.contains("aarch64"))
        } else if platform == "linux-x86_64" {
            name.ends_with(".appimage") || name.ends_with(".AppImage") || name.ends_with(".deb")
        } else {
            false
        }
    })
}

